types.go

```diff
@@ -15,33 +15,34 @@ import (
 
 const defaultGradientAbsTol = 1e-6
 
-// EvaluationType is used by the optimizer to specify information needed
-// from the objective function.
+// EvaluationType is used by the optimizer to specify information needed from
+// the objective function. The types can be composed together using binary or
+// operator.
 type EvaluationType int
 
 const (
-	NoEvaluation EvaluationType = iota
-	FuncEvaluation
-	GradEvaluation
-	HessEvaluation
-	FuncGradEvaluation
-	FuncGradHessEvaluation
+	NoEvaluation           EvaluationType = 0
```

Comment:
Don't write it with `+`. If this is done and you typo a repeated flag, then you get a shift.

Also, @vladimir-ch, you can use `1<<iota` for the second line and elide the next two.